clippy_lints/src/fn_param_redef_as_mutable.rs

```diff
@@ -0,0 +1,96 @@
+use rustc_lint::{EarlyLintPass, EarlyContext};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use syntax::ast::*;
+use syntax::visit::FnKind;
+use rustc_span::Span;
+use rustc_errors::DiagnosticBuilder;
+use crate::utils::{span_lint_and_then, multispan_sugg};
+use if_chain::if_chain;
+
+declare_clippy_lint! {
+    /// **What it does:** checks if any fn parameters have been assigned to a local mutable
+    /// variable.
+    ///
+    /// **Why is this bad?** reduces the complexity of the code by removing a redundant local
+    /// mutable variable.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// fn f(a: Vec<bool>) {
+    ///     let mut a = a;
+    ///     // rest of code
+    /// }
+    /// ```
+    /// 
+    /// could be defined as 
+    ///
+    /// ```rust
+    /// fn f(mut a: Vec<bool>) {
+    ///     // rest of code
+    /// }
+    /// ```
+    pub FN_PARAM_REDEF_AS_MUTABLE,
```

Comment:
bikeshedding: I think we usually call them variable "bindings" rather than "definitions". So, this can be named `FN_PARAM_REBIND_AS_MUT`.

Reply:
Sweet, I'll rename it cause that sounds a lot better.